### PR TITLE
add multiBodyMacro

### PR DESCRIPTION
**Pull request by krux02**
Well there is the need to pass multiple bodies to macros. There is a pattern that I already use for quite some time successfully. Because there was request for it, I wrote a helper macro for it. This is the first version that I got working, not something that should be merged already. This PR does not have it's RFC yet. Things that are missing are more extensive tests.

For the examples code here:
```nim
macro myMacro(body1,body2: untyped): untyped {.multiBodyMacro.} =
  echo body1.lispRepr # (StmtList (Command (Ident "echo") (IntLit 1)))
  echo body2.lispRepr # (StmtList (Command (Ident "echo") (IntLit 2)))

myMacro:
  body1:
    echo 1
  body2:
    echo 2
```
This is the generated code:
```nim
macro myMacro(metabody: untyped): untyped =
  var body1, body2: NimNode
  for body in metabody:
    body.expectKind nnkCall
    body.expectLen 2
    if eqIdent(body[0], "body1"):
      body1 = body[1]
    elif eqIdent(body[0], "body2"):
      body2 = body[1]
    else:
      error("unexpectecd section", body[0])
  echo body1.lispRepr
  echo body2.lispRepr
```

### CODE REVIEW

**Comment by krux02** [comment]
Personally I wouldn't want to use ``multiBodyMacro`` in my own code and rather have the expanded macro written down in my own code. Simply because that provides more flexibility, for example more arguments that not not a section, or code that isn't part of a section but still contains logic. After all this macro doesn't generate a ton of code that would be either tedious or error prone to write manually, it simply automates a pattern that turned out to be very useful for me. But I think the inclusion of this in the standard library is valuable, because it is a useful and well written example of a pragma expression macro. We don't have these yet, as far as I know.

Maybe it is a good idea to put this code not in macros.nim, but instead directly in the tutorial for macros. But I would need some feedback from you. This PR is a new feature and new features should go through a review process. So feel free to see this PR as an RFC and post your opinions, thoughts and ideas.

**Comment by Vindaar** [comment]
I think this would be a very nice addition to the macros tutorial. It's a simple example for something useful that's also easy to understand.

Still I think it's a good idea to have in the macros module as well though. For two reasons:
1. the resulting code is simple enough that I don't have to worry about introducing any weird complications. 
2. for simple macros this makes the code a lot shorter. Sure if I'm writing a complex macro / a DSL or whatever, I'm going to do this stuff by hand anyways. But for simple helper macros this is great (assuming the `do` notation is removed anyways). For instance in my [shell](https://github.com/Vindaar/shell) dsl I have a macro for my unittests, which essentially boils down to the following with the multiBody macro (see [here](https://github.com/Vindaar/shell/blob/master/shell.nim#L376-L390) for the actual one):
```nim
macro checkShell(body1, body2: untyped): untyped {.multiBodyMacro.} =
  let shellCmds = genShellCmds(body1)
  result = quote do:
    check `shellCmds` == `body2`

checkShell:
  body1:
    ./test ("--out="$outname)
  body2:
    &"./test --out={outname}"
```
At the moment I rely on the do notation to not having to extract two bodies from one argument.

**Comment by Araq** [comment]
Yes, please add it to the tutorial instead. We don't need more bloat in ``macros.nim``...

**Comment by mratsim** [comment]
I don't think it's bloat, now maybe it should be insugar instead of macros.

Macros as pragmas are very useful in theory but in practice I'm always losing 10~30min to rediscover the syntax and how parameters are supposed to be passed (see: https://forum.nim-lang.org/t/5450#34078). Some tutorial or cookbook/recipes would be great but an off-the-shelf solution is very useful as well.

File: doc/tut3.rst [hunk 1]
```diff
@@ -289,6 +289,132 @@ used to get this output.
...
+
+Advanced Macro to allow Multi Body Macros
+-----------------------------------------
+
+Sometimes it is necessary to pass multiple blocks of code to a
+macro. While it is possible to pass an arbitrary amount of expression
+to a macro, the only argument that truly allows multi line blocks of
+code is the last argument. But we can introduce a macro that allows us
+to interpret individual sub-trees of the last parameter as if they were
+different arguments. Let me provide you with an example of what I
```
Comment (Araq) [inline doc/tut3.rst]: "Let me ... " -> "For example:"

File: doc/tut3.rst [hunk 2]
```diff
@@ -289,6 +289,132 @@ used to get this output.
...
+
+The call of ``myMacro`` has actually just one argument. But this one
+argument contains two calls ``body1:...`` and ``body2:...``. Therefore
+the code above would normally not compile, because ``myMacro`` needs
+two arguments. But the pragma expression ``{.multiBodyMacro.}`` is
+itself a macro that transforms the macro definition of ``myMacro``
+into a definition with just one argument, and a processing block that
+initializes two variables, ``body1`` and ``body2`` with their
+corresponding sub-tree in the single argument. Here is the macro that
+does this processing.
```
Comment (Araq) [inline doc/tut3.rst]: Too much code following, better leave it out.

**Pull request closed without merging**